feat: add Kotlin 1.7.20-Beta EAP release details (#3076)
JetBrains/kotlin-web-site: docs/topics/whatsnew-eap.md

>
{type="note"}

The Kotlin 1.7.20-Beta release is out! Here are some highlights from this preview version of Kotlin:

* [The new Kotlin K2 compiler supports Lombok, `all-open`, Parcelize, and other compiler plugins](#support-for-kotlin-k2-compiler-plugins)
* [We introduced the preview of the `..<` operator for creating open-ended ranges](#preview-of-the-operator-for-creating-open-ended-ranges)
* [The new Kotlin/Native memory manager enabled by default](#the-new-kotlin-native-memory-manager-enabled-by-default)
* [We introduced a new experimental feature for JVM: inline classes with a generic underlying type](#generic-inline-classes)

## Support for Kotlin K2 compiler plugins

The Kotlin team continues to stabilize the K2 compiler.
K2 is still in Alpha (as [announced in the Kotlin 1.7.0 release](whatsnew17.md#new-kotlin-k2-compiler-for-the-jvm-in-alpha)), but now it supports several compiler plugins.
You can follow [this YouTrack issue](https://youtrack.jetbrains.com/issue/KT-52604) to get updates from the Kotlin team on the new compiler.

Starting with this preview version, the Kotlin K2 compiler supports the following plugins:

> * [All-open](all-open-plugin.md)

Reviewer: ```suggestion
* [`all-open`](all-open-plugin.md)
```

> * [No-Arg](no-arg-plugin.md)

Reviewer: ```suggestion
* [`no-arg`](no-arg-plugin.md)
```

* [SAM with receiver](sam-with-receiver-plugin.md)
* [Lombok](lombok.md)
* Parcelize

> * atomicfu

Reviewer: ```suggestion
* AtomicFU
```

> * jvm-abi-gen

Reviewer: ```suggestion
* `jvm-abi-gen`
```

The new memory manager finally paves the way to [promote Kotlin Multiplatform Mobile to Beta](https://blog.jetbrains.com/kotlin/2022/05/kotlin-multiplatform-mobile-beta-roadmap-update/).

The new memory manager also supports the compiler cache that makes compilation times comparable to previous releases.
For more on the benefits of the new memory manager, see our original [blog post](https://blog.jetbrains.com/kotlin/2021/08/try-the-new-kotlin-native-memory-manager-development-preview/?_ga=2.78056339.62647407.1657550793-1365835714.1628072746&_gl=1*1dslhfh*_ga*MTM2NTgzNTcxNC4xNjI4MDcyNzQ2*_ga_9J976DJZ68*MTY1Nzg2NDY5NS4yMDIuMS4xNjU3ODY1MTU0LjA.) for the preview version.
You can find more technical details in the [migration instructions on GitHub](https://github.com/JetBrains/kotlin/blob/master/kotlin-native/NEW_MM.md).

### Configuration and setup

Starting with Kotlin 1.7.20, the new memory manager is the default. Not much additional setup is required.

If you've already turned it on manually, you can remove the `kotlin.native.binary.memoryModel=experimental` option from your `gradle.properties` or `binaryOptions["memoryModel"] = "experimental"` from the `build.gradle(.kts)` file.

> If it’s necessary, you can switch back to the legacy memory manager with the `kotlin.native.binary.memoryModel=strict` option in your `gradle.properties`.

Reviewer: ```suggestion
If it's necessary, you can switch back to the legacy memory manager with the `kotlin.native.binary.memoryModel=strict` option in your `gradle.properties`.
```

However, compiler cache support is not available for the legacy memory manager anymore, so compilation times might become worse.

#### Freezing

In the new memory manager, freezing is deprecated. Don't use it unless you need your code to work with the legacy manager (where freezing is still required).
This may be helpful for library authors that need to maintain support for the legacy memory manager or developers who want to have a fallback if they encounter issues with the new memory manager.

In such cases, you can temporarily support code for both new and legacy memory managers. To ignore deprecation warnings, do one of the following:

* Annotate usages of the deprecated API with `@OptIn(FreezingIsDeprecated::class)`.
* Apply `languageSettings.optIn("kotlin.native.FreezingIsDeprecated")` to all the Kotlin source sets in Gradle.
* Pass the compiler flag `-opt-in=kotlin.native.FreezingIsDeprecated`.

#### Calling Kotlin suspending functions from Swift/Objective-C

The new memory manager still has a restriction on calling Kotlin `suspend` functions from Swift and Objective-C from threads other than the main one, but you can lift it with a new Gradle option.

> This restriction was originally introduced in the legacy memory manager due to cases where the code dispatched a continuation to be resumed on the original thread. If this thread didn’t have a supported event loop, the task would never run, and the coroutine would never be resumed.

Reviewer: ```suggestion
This restriction was originally introduced in the legacy memory manager due to cases where the code dispatched a continuation to be resumed on the original thread. If this thread didn't have a supported event loop, the task would never run, and the coroutine would never be resumed.
```

## Generic inline classes

> Generic inline classes is an [Experimental](components-stability.md#stability-levels-explained) feature.
> It may be dropped or changed at any time. Opt-in is required (see details below), and you should use it only for evaluation purposes.
> We would appreciate your feedback on it in [YouTrack](https://youtrack.jetbrains.com/issue/KT-52994).
>
{type="warning"}

Kotlin 1.7.20-Beta allows the underlying type of JVM inline classes to be a type parameter.
The compiler maps it to `Any?` or, generally, to the upper bound of the type parameter.

Consider the following example:

```kotlin
@JvmInline
value class IC<T>(val a: T)

fun foo(s: IC<String>) {} // compiler generates fun foo-<hash>(s: Any?)
```

The function accepts the inline class as a parameter. The parameter is mapped to the upper bound and not to the type argument.

To enable this feature, use the `-language-version 1.8` compiler option.

## How to update to the Kotlin 1.7.20-Beta

Install Kotlin 1.7.0-Beta in any of the following ways:

* If you use the _Early Access Preview_ update channel, the IDE will suggest automatically updating to 1.7.20-Beta as soon as it becomes available.

> * If you use the _Stable_ update channel, you can change the channel to _Early Access Preview_ at any time by selecting **Tools** | **Kotlin** | **Configure Kotlin Plugin Updates** in your IDE. You’ll then be able to install the latest preview release. Check out [these instructions](install-eap-plugin.md) for details.

Reviewer: ```suggestion
* If you use the _Stable_ update channel, you can change the channel to _Early Access Preview_ at any time by selecting **Tools** | **Kotlin** | **Configure Kotlin Plugin Updates** in your IDE. You'll then be able to install the latest preview release. Check out [these instructions](install-eap-plugin.md) for details.
```

> Once you’ve installed 1.7.20-Beta, don’t forget to [change the Kotlin version](configure-build-for-eap.md) to 1.7.20-Beta in your build scripts.

Reviewer: ```suggestion
Once you've installed 1.7.20-Beta, don't forget to [change the Kotlin version](configure-build-for-eap.md) to 1.7.20-Beta in your build scripts.
```
